### PR TITLE
compose: Add types to `useDialog` and `useFocusOnMount`

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
This PR finally finishes adding types to `compose` by adding types to `useFocusOnMount` (which was a simple small change), fixing the types for `useConstrainedTabbing` (not sure how that got through review but it was completely broken, just a one line change to fix it though) and adding types to `useDialog`.

Part of #18838

This also updates the CHANGELOG and adds the `types` directive in the `compose/package.json`.

This should be merged after #32674 

## How has this been tested?
Type checks pass.

## Types of changes
New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/compose/tsconfig.json
Comment (ciampo) [inline packages/compose/tsconfig.json]: Yes!

**Review by ciampo — COMMENTED** [review]
Everything seems good so far, but probably better to wait for #32674 to be merged and rebase, since this PR also introduces changes to the whole `compose` package assuming that it is fully typed.

**Comment by sarayourfriend** [comment]
@ciampo Ready for another review after the previous PR merged!

**Review by ciampo — APPROVED** [review]
LGTM 🚀 

Great job typing the whole package!

**Pull request merged**
